## **Exercicios 01 - Cálculo Numérico**

**Wanderson Faustino Patricio**


In [13]:
# Antes das soluções vamos importar as bibliotecas necessárias
import numpy as np
import numpy.linalg as la
from math import *

In [2]:
def badSolver(A : np.array, b : np.array) -> np.array:
    """Solves the system Ax = b
    A and b must be of the np.array
    """
    At = A.transpose()
    AtA = At.dot(A)
    Atb = At.dot(b)

    return la.solve(AtA, Atb)

In [32]:
def linAlgSolver(A : np.array, b : np.array) -> np.array:
    """Solves the system Ax = b with the library np.linalg
    A and b must be of the np.array
    """
    return la.lstsq(A, b)[0]

**Questão 01**

Resolva o sistema abaixo usando o metodo dos mínimos quadrados.

$$
\left[
\begin{array}{cc}
1 & 1 \\
2 & 1 \\
3 & 1
\end{array}
\right]
\left[
\begin{array}{c}
x_0 \\
x_1
\end{array}
\right] =
\left[
\begin{array}{c}
1 \\
2 \\
0
\end{array}
\right]
$$

***Solução Analítica***

Multiplicando pela transposta da primeira matriz em ambos os lados

$$
\left[
\begin{array}{ccc}
1 & 2 & 3 \\
1 & 1 & 1
\end{array}
\right]
\left[
\begin{array}{cc}
1 & 1 \\
2 & 1 \\
3 & 1
\end{array}
\right]
\left[
\begin{array}{c}
x_0 \\
x_1
\end{array}
\right] =
\left[
\begin{array}{ccc}
1 & 2 & 3 \\
1 & 1 & 1
\end{array}
\right]
\left[
\begin{array}{c}
1 \\
2 \\
0
\end{array}
\right]
$$

$$
\left[
\begin{array}{cc}
14 & 6 \\
6 & 3
\end{array}
\right]
\left[
\begin{array}{c}
x_0 \\
x_1
\end{array}
\right]=
\left[
\begin{array}{c}
5 \\
3
\end{array}
\right]
$$

A solução será

$$
\left[
\begin{array}{c}
x_o \\
x_1
\end{array}
\right]=
\left[
\begin{array}{c}
-\dfrac{1}{2} \\
2
\end{array}
\right]
$$

***Solução Numérica***

In [ ]:
A = np.array([[1, 1],[2,1],[3,1]])
b = np.array([1,2,0])

badSolution = badSolver(A, b)
linalgSolution = linAlgSolver(A, b)

In [7]:
print("Bad Solution:")
badSolution

Bad Solution:


array([-0.5,  2. ])

In [9]:
print("linAlg Solution:")
linalgSolution

linAlg Solution:


array([-0.5,  2. ])

**Questão 02**

a) Aplicando os valores



*   $f(0)=a+b=26$
*   $f(6)=a+c=21$
*   $f(12)=a-b=34$
*   $f(18)=a-c=30$

teremos o seguinte sistema

$$
\left[
\begin{array}{ccc}
1 & 1 & 0 \\
1 & 0 & 1 \\
1 & -1 & 0 \\
1 & 0 & -1 \\
\end{array}
\right]
\left[
\begin{array}{c}
a \\
b \\
c
\end{array}
\right]=
\left[
\begin{array}{c}
26 \\
21 \\
34 \\
30
\end{array}
\right]
$$

b) Seguindo a mesma ideia do item anterior

$$
\left[
\begin{array}{cccc}
1 & 1 & 1 & 1 \\
1 & 0 & -1 & 0 \\
0 & 1 & 0 & -1
\end{array}
\right]
\left[
\begin{array}{ccc}
1 & 1 & 0 \\
1 & 0 & 1 \\
1 & -1 & 0 \\
1 & 0 & -1 \\
\end{array}
\right]
\left[
\begin{array}{c}
a \\
b \\
c
\end{array}
\right]=
\left[
\begin{array}{cccc}
1 & 1 & 1 & 1 \\
1 & 0 & -1 & 0 \\
0 & 1 & 0 & -1
\end{array}
\right]
\left[
\begin{array}{c}
26 \\
21 \\
34 \\
30
\end{array}
\right]
$$

$$
\left[
\begin{array}{ccc}
4 & 0 &  0 \\
0 & 2 & 0 \\
0 & 0 & 2
\end{array}
\right]\left[
\begin{array}{c}
a \\
b \\
c
\end{array}
\right]=
\left[
\begin{array}{c}
111 \\
-8\\
-9
\end{array}
\right]
$$

O que nos dá a solução

$$
\left[
\begin{array}{c}
a \\
b \\
c
\end{array}
\right]=
\left[
\begin{array}{c}
\dfrac{111}{4} \\
-4\\
-\dfrac{9}{2}
\end{array}
\right]=
\left[
\begin{array}{c}
27.25 \\
-4\\
-4.5
\end{array}
\right]
$$

In [ ]:
A = np.array([[1, 1, 0],[1, 0, 1],[1, -1, 0],[1,0,-1]])
b = np.array([26, 21, 34, 30])

badSolution = badSolver(A, b)
linalgSolution = linAlgSolver(A, b)

In [11]:
print("Bad Solution:")
badSolution

Bad Solution:


array([27.75, -4.  , -4.5 ])

In [12]:
print("linAlg Solution:")
linalgSolution

linAlg Solution:


array([27.75, -4.  , -4.5 ])

c) $T(14)=27.75 - 4\cdot \cos{\left(\dfrac{7\pi}{6}\right)}-7.5\cdot \sin{\left(\dfrac{7\pi}{6}\right)}=33.464^{\circ}C$

In [16]:
linalgSolution[0] + linalgSolution[1] * cos((7*pi)/6) + linalgSolution[2] * sin((7*pi)/6)

33.46410161513776

**Questão 03**

a) Aplicando os valores



*   $f(0)=a+c=0$
*   $f(0.5)=a\cdot \sqrt{e}+c=0.57$
*   $f(1)=a\cdot e + c=1.46$
*   $f(2)=a\cdot e^2+ c=5.05$

teremos o seguinte sistema

$$
\left[
\begin{array}{cc}
1 & 1 \\
\sqrt{e} & 1  \\
e & 1 \\
e^2 & 1 \\
\end{array}
\right]
\left[
\begin{array}{c}
a \\
c
\end{array}
\right]=
\left[
\begin{array}{c}
0 \\
0.57 \\
1.46 \\
5.05
\end{array}
\right]
$$

Fazendo mínimos quadrados

$$
\left[
\begin{array}{cccc}
1 & \sqrt{e} & e & e^2 \\
1&1&1&1
\end{array}
\right]
\left[
\begin{array}{cc}
1 & 1 \\
\sqrt{e} & 1  \\
e & 1 \\
e^2 & 1 \\
\end{array}
\right]
\left[
\begin{array}{c}
a \\
c
\end{array}
\right]=
\left[
\begin{array}{cccc}
1 & \sqrt{e} & e & e^2 \\
1&1&1&1
\end{array}
\right]
\left[
\begin{array}{c}
0 \\
0.57 \\
1.46 \\
5.05
\end{array}
\right]
$$

$$
\left[
\begin{array}{cc}
65.704 & 12.756 \\
12.756 & 4 \\
\end{array}
\right]
\left[
\begin{array}{c}
a \\
c
\end{array}
\right]=
\left[
\begin{array}{c}
42.223 \\
7.08
\end{array}
\right]
$$

A solução será

$$
\left[
\begin{array}{c}
a \\
c
\end{array}
\right]=\left[
\begin{array}{c}
0.785 \\
-0.733
\end{array}
\right]
$$

In [ ]:
A = np.array([[1, 1],[sqrt(e), 1],[e, 1],[e**2,1]])
b = np.array([0,0.57,1.46,5.05])

badSolution = badSolver(A, b)
linalgSolution = linAlgSolver(A, b)

In [29]:
print("Bad Solution:")
badSolution

Bad Solution:


array([ 0.78497536, -0.73329805])

In [30]:
print("linAlg Solution:")
linalgSolution

linAlg Solution:


array([ 0.78497536, -0.73329805])

b) Aplicando os valores de $a$ e $c$ para achar b:

*   $f(0)=0.785 \cdot e^{b\cdot 0}-0.733=0$
*   $f(0.5)=0.785 \cdot e^{b\cdot 0.5}-0.733=0.57$
*   $f(1)=0.785 \cdot e^{b\cdot 1}-0.733=1.46$
*   $f(2)=0.785 \cdot e^{b\cdot 2}-0.733=5.05$

Organizando
*   $e^{b\cdot 0}=0.934$
*   $e^{b\cdot 0.5}=1.660$
*   $e^{b\cdot 1}=2.794$
*   $e^{b\cdot 2}=7.367$

Encontramos portanto o sistema

$$
\left[
\begin{array}{c}
0 \\
0.5 \\
1 \\
2
\end{array}
\right]
\left[
\begin{array}{c}
b
\end{array}
\right]=
\left[
\begin{array}{c}
\ln{(0.934)} \\
\ln{(1.660)} \\
\ln{(2.794)} \\
\ln{(7.367)} \\
\end{array}
\right]
$$

$$
\left[
\begin{array}{cccc}
0 & 0.5 &1 &2
\end{array}
\right]
\left[
\begin{array}{c}
0 \\
0.5 \\
1 \\
2
\end{array}
\right]
\left[
\begin{array}{c}
b
\end{array}
\right]=
\left[
\begin{array}{cccc}
0 & 0.5 &1 &2
\end{array}
\right]
\left[
\begin{array}{c}
\ln{(0.934)} \\
\ln{(1.660)} \\
\ln{(2.794)} \\
\ln{(7.367)} \\
\end{array}
\right]
$$

$$
\left[
\begin{array}{c}
5.25
\end{array}
\right]
\left[
\begin{array}{}
b
\end{array}
\right]=
\left[
\begin{array}{}
5,275
\end{array}
\right]
$$

Portanto:

$$b=1.00476$$

In [ ]:
A = np.array([[0], [0.5], [1], [2]])
b = np.log(np.array([0.934, 1.66, 2.794, 7.367]))

badSolution = badSolver(A, b)
linalgSolution = linAlgSolver(A, b)

In [42]:
print("Bad Solution:")
badSolution

array([1.00474366])

In [43]:
print("linAlg Solution:")
linalgSolution

linAlg Solution:


array([1.00474366])

c) Faremos agora o cálculo do coeficiente de correlação dos dois modelos.

In [60]:
def mean(lista):
    sum = 0
    for item in lista:
        sum += item

    return sum/lista.size

def R(y_real, y_pred):
    n = y_real.size
    y_mean = mean(y_real)
    SQtot = 0
    for index in range(n):
        SQtot += (y_real[index]- y_mean)**2

    SQres = 0
    for index in range(n):
        SQres += (y_real[index]- y_pred[index])**2

    coeff = 1 - SQres/SQtot
    return coeff

In [61]:
y_real = np.array([0,0.57,1.46,5.05])
x = np.array([0, 0.5, 1, 2])
y_pred1 = 0.785 * np.exp(x) - 0.733
y_pred2 = 0.785 * np.exp(1.00474366*x) - 0.733

R1 = R(y_real, y_pred1)
R2 = R(y_real, y_pred2)

print(f'Coeficiente da abordagem 1 : {R1}')
print(f'Coeficiente da abordagem 2 : {R2}')

Coeficiente da abordagem 1 : 0.9995733383453307
Coeficiente da abordagem 2 : 0.999324385247391


A primeira abordagem gerou um coeficiente de correlação maior, ou seja, foi mais preciso

**Questão 04**

a) Trabalhando um pouco a equação, podemos chegar ao seguinte resultado

$$\sqrt{y} = \dfrac{a}{b}\cdot \dfrac{1}{\sqrt{x}} + \dfrac{1}{b}$$

Ou seja, uma equação do tipo

$$u = m\cdot v + n$$

com $u=\sqrt{y}$, $v=\dfrac{1}{\sqrt{x}}$, $m=\dfrac{a}{b}$  e  $n=\dfrac{1}{b}$.

b) Reescrevendo a tabela

$$
u = \left[3.225, 2.408, 1.817, 1.549, 1.414\right]
$$

$$
v = \left[1.414, 1, 0.707, 0.577, 0.5\right]
$$

Teremos o seguinte sistema

$$
\left[
\begin{array}{cc}
1.414 & 1 \\
1 & 1 \\
0.707 & 1 \\
0.577 & 1 \\
0.5 & 1
\end{array}
\right]
\left[
\begin{array}{c}
m \\
n
\end{array}
\right]=
\left[
\begin{array}{c}
3.225\\
2.408\\
1.817\\
1.549\\
1.414
\end{array}
\right]
$$

Aplicando o método dos mínimos quadrados

$$
\left[
\begin{array}{cccc}
1.414 & 1 & 0.707 & 0.577 & 0.5 \\
1 & 1 & 1 & 1 & 1
\end{array}
\right]
\left[
\begin{array}{cc}
1.414 & 1 \\
1 & 1 \\
0.707 & 1 \\
0.577 & 1 \\
0.5 & 1
\end{array}
\right]
\left[
\begin{array}{c}
m \\
n
\end{array}
\right]=
\left[
\begin{array}{cccc}
1.414 & 1 & 0.707 & 0.577 & 0.5 \\
1 & 1 & 1 & 1 & 1
\end{array}
\right]
\left[
\begin{array}{c}
3.225\\
2.408\\
1.817\\
1.549\\
1.414
\end{array}
\right]
$$

$$\left[
\begin{array}{cc}
4.082 & 4.198 \\
4.198 & 5
\end{array}
\right]
\left[
\begin{array}{c}
m \\
n
\end{array}
\right]=
\left[
\begin{array}{c}
9.854 \\
10.413
\end{array}
\right]
$$

O que nos conduz à solução
$$
\left[
\begin{array}{c}
m \\
n
\end{array}
\right]=
\left[
\begin{array}{c}
1.992 \\
0.410
\end{array}
\right]
$$

In [ ]:
A = np.array([[1.414, 1], [1, 1], [0.707, 1], [0.577, 1], [0.5, 1]])
b = np.array([3.225, 2.408, 1.817, 1.549, 1.414])

badSolution = badSolver(A, b)
linalgSolution = linAlgSolver(A, b)

In [67]:
print("Bad Solution:")
badSolution

Bad Solution:


array([1.99232476, 0.40984413])

In [68]:
print("linAlg Solution:")
linalgSolution

linAlg Solution:


array([1.99232476, 0.40984413])

Temos

$$n=\dfrac{1}{b}=0.410 \Rightarrow b = 2.439$$
$$m = \dfrac{a}{b} ⇒ 1.992 = \dfrac{a}{2.439} ⇒ a=4.858$$